services/users.go

```diff
@@ -0,0 +1,302 @@
+package services
+
+import (
+	"golang.org/x/crypto/bcrypt"
+	"gorm.io/gorm"
+
+	"fmt"
+	"net/http"
+
+	"mini-social-network/constants"
+	"mini-social-network/errors"
+	"mini-social-network/models"
+	"mini-social-network/serializers"
+	"mini-social-network/utils"
+)
+
+func (s *Service) GetUserByID(userID uint) (*serializers.GetUserResponse, *errors.APIError) {
+	var user models.User
+	if err := s.DB.Preload("OfficeDetails").Preload("ResidentialDetails").
+		Where("id = ? AND deleted_at IS NULL", userID).
+		Scan(&user).Error; err != nil {
+		if err == gorm.ErrRecordNotFound {
+			return nil, errors.NewAPIError(constants.ErrUserNotFound, http.StatusNotFound)
+		}
+		return nil, errors.NewAPIError(constants.ErrInternalServerError, http.StatusInternalServerError)
+	}
+
+	userResponse := serializers.SerializeResponse(user, user.ResidentialDetails, user.OfficeDetails)
+	response := serializers.SerializeGetUserResponse(user, userResponse)
+
+	return &response, nil
+}
+
+func (s *Service) ListUsers() ([]serializers.ListUserResponse, *errors.APIError) {
+	var users []models.User
+
+	if err := s.DB.Select("id, email").Find(&users).Error; err != nil {
+		return nil, errors.NewAPIError(constants.ErrFailedToRetrieveUser, http.StatusInternalServerError)
+	}
+
+	response := serializers.SerializeListUser(users)
+
+	return response, nil
+}
+
+func (s *Service) UpdatePasswordByID(userID uint, req serializers.PasswordRequest) *errors.APIError {
+	var user models.User
+	if err := s.DB.Where("id = ?", userID).First(&user).Error; err != nil {
+		return errors.NewAPIError(constants.ErrUserNotFound, http.StatusNotFound)
+	}
+
+	if err := bcrypt.CompareHashAndPassword([]byte(user.Password), []byte(req.OldPassword)); err != nil {
+		return errors.NewAPIError(constants.ErrInvalidOldPassword, http.StatusBadRequest)
+	}
+
+	newPasswordHash, err := utils.HashPassword(req.NewPassword)
+	if err != nil {
+		return errors.NewAPIError(constants.ErrFailedToHashPassword, http.StatusInternalServerError)
+	}
+
+	if err := s.DB.Model(&models.User{}).Where("id = ?", userID).Update("password", newPasswordHash).Error; err != nil {
+		return errors.NewAPIError(constants.ErrFailedToUpdatePass, http.StatusInternalServerError)
+	}
+
+	return nil
+}
+
+func (s *Service) DeleteUserByID(userID uint) (*serializers.DeleteUserResponse, *errors.APIError) {
+	tx := s.DB.Begin()
+	var user models.User
+
+	defer func() {
+		if r := recover(); r != nil {
+			tx.Rollback()
+		}
+	}()
+
+	if err := tx.Preload("OfficeDetails").Preload("ResidentialDetails").First(&user, "id = ?", userID).Error; err != nil {
+		if err == gorm.ErrRecordNotFound {
+			tx.Rollback()
+			return nil, errors.NewAPIError(constants.ErrUserNotFound, http.StatusNotFound)
+		}
+		tx.Rollback()
+		return nil, errors.NewAPIError(constants.ErrFailedToRetrieveUser, http.StatusInternalServerError)
+	}
```

Comment:
```suggestion
	if err := tx.Preload("OfficeDetails").Preload("ResidentialDetails").First(&user, "id = ?", userID).Error; err != nil {
	       tx.Rollback()
		if err == gorm.ErrRecordNotFound {
			return nil, errors.NewAPIError(constants.ErrUserNotFound, http.StatusNotFound)
		}
		return nil, errors.NewAPIError(constants.ErrInternalServerError, http.StatusInternalServerError)
	}
```
Also update error message to ErrInternalServerError in case of 500